bin/gather_changelog/gather_changelog.sh

```diff
@@ -0,0 +1,330 @@
+    container_name=$1
+    function_string='
+    function git_commit_dict() {
+    # Define an empty dictionary
+    declare -A git_dict
+
+    # Loop through all directories in the given path
+    for dir in "$1"/*/; do
+        cd $dir
+        # Get the repo name by removing the path and trailing slash
+        repo_name="$(basename "${dir%/}")"
+        remote_url=$(git config --get remote.origin.url)
+        repo_name=$(basename "$remote_url")
+        if [[ "$remote_url" == *"github.com:shadow-robot"* ]]; then
+            # Get the git commit ID and store it in the dictionary
+            git_dict[$repo_name]="$(git -C "$dir" rev-parse HEAD)"
+        fi
+    done
+
+    for key in "${!git_dict[@]}"; do
+        echo "$key:${git_dict[$key]}" >> /home/user/git_commit_dict.txt
+    done
+    }
+
+    git config --global --add safe.directory "*"
+    git_commit_dict $1
+    '
+    docker exec -i $container_name bash -c "echo -e '$function_string' > /home/user/git_commit_dict.sh"
+    docker exec -i $container_name chmod +x /home/user/git_commit_dict.sh
+}
+
+# This function is used to start a docker container, then write a script to the container to gather all repository's
+# information (repo name and hash) given a path, it then takes the file it writes the information too and copies it
+# to the local machine to be read.
+function get_all_github_dependencies() {
+    container_name=$1
+    image_name=$2
+
+    docker run -itd --name $container_name $image_name
+    write_git_commit_dict_function_to_container $container_name
+    docker exec -i $container_name /bin/bash -c 'home/user/git_commit_dict.sh /home/user/projects/shadow_robot/base/src'
+    docker exec -i $container_name /bin/bash -c 'home/user/git_commit_dict.sh /home/user/projects/shadow_robot/base_deps/src'
+    docker cp $container_name:home/user/git_commit_dict.txt /tmp/$container_name.txt
+    docker stop $container_name
+}
+
+# This function takes in a path to a file and will go through the file line by line and creates a dictionarys.
+# The file must be in the format `key:value` with one entry per line.
+function file_to_dict() {
+    declare -g -A file_dict=()
+
+    while read line; do
+        key=$(echo $line | cut -d ":" -f1)
+        data=$(echo $line | cut -d ":" -f2)
+        file_dict[$key]=$data
+    done < $1
+}
+
+# This function takes in two dictionarys and compares them both. We use this as a way of determining which repoitories
+# have been added, removed and updated. It sets 4 global variables:
+# updated_repos - This is the dict that contains a key of the repo name and the value of the previous containers hash
+# updated_repos_recent - Has the same key as above but has the current containers hash
+# added_repos - This has the same key as above and has the current containr hash
+# removed_repos - this is a list of all the repositories that have been removed between dict 1 and dict 2.
+function compare_dicts() {
+    local -n array1=$1
```

Comment:
Can you add a comment on which array should contain the dictionary with contents of "most recent image"? It seems that would be `array1`?